File: packages/python/src/erpc_sdk/transport.py
"""Async HTTP transports with explicit timeout, cancellation, and redaction."""

from __future__ import annotations

import asyncio
import itertools
from collections.abc import AsyncIterator, Awaitable, Iterable, Mapping, Sequence
from contextlib import asynccontextmanager
from typing import TypeVar, cast
from urllib.parse import urlsplit, urlunsplit

import httpx

from .config import endpoint_with_path
from .errors import (
    ErpcAbortedError,
    ErpcConfigError,
    ErpcHttpError,
    ErpcInvalidResponseError,
    ErpcJsonRpcError,
    ErpcNotConfiguredError,
    ErpcTimeoutError,
    ErpcTransportError,
    redact_text,
    redact_value,
)
from .types import (
    DEFAULT_REQUEST_OPTIONS,
    JsonRpcId,
    JsonRpcParams,
    JsonValue,
    RequestOptions,
    RpcBatchCall,
)

T = TypeVar("T")
QueryValue = bool | float | int | Sequence[str] | str | None


def _response_json(response: httpx.Response) -> object:
    """Parse a response without retaining parser context or response bodies."""
    parse_error: ErpcInvalidResponseError | None = None
    try:
        value = response.json()
    except ValueError:
        parse_error = ErpcInvalidResponseError("ERPC returned malformed JSON")
    if parse_error is not None:
        # Raising after leaving the native parser's exception handler keeps
        # its body and context out of the public SDK error.
        raise parse_error
    return value


def _direct_headers(headers: Mapping[str, str]) -> httpx.Headers:
    """Build direct headers without retaining native construction errors."""
    construction_error: ErpcConfigError | None = None
    try:
        result = httpx.Headers(headers)
    except (TypeError, UnicodeError, ValueError):
        construction_error = ErpcConfigError(
            "direct endpoint headers must contain valid HTTP values"
        )
    if construction_error is not None:
        raise construction_error
    result["accept"] = "application/json"
    result["content-type"] = "application/json"
    return result


def _direct_request(endpoint: str, headers: httpx.Headers, body: object) -> httpx.Request:
    """Build a direct request while discarding URL/body construction details."""
    construction_error: ErpcTransportError | None = None
    try:
        request = httpx.Request("POST", endpoint, headers=headers, json=body)
    except (TypeError, UnicodeError, ValueError, httpx.InvalidURL):
        construction_error = ErpcTransportError("Unable to construct direct HTTP request")
    if construction_error is not None:
        raise construction_error
    return request


async def _wait(awaitable: Awaitable[T], timeout: float, options: RequestOptions) -> T:
    request_task = asyncio.ensure_future(awaitable)
    cancel_task: asyncio.Task[bool] | None = None
    try:
        async with asyncio.timeout(timeout):
            if options.cancel_event is None:
                return await request_task
            if options.cancel_event.is_set():
                request_task.cancel()
                raise ErpcAbortedError
            cancel_task = asyncio.create_task(options.cancel_event.wait())
            done, _ = await asyncio.wait(
                {request_task, cancel_task}, return_when=asyncio.FIRST_COMPLETED
            )
            if cancel_task in done:
                request_task.cancel()
                await asyncio.gather(request_task, return_exceptions=True)
                raise ErpcAbortedError
            cancel_task.cancel()
            return await request_task
    except TimeoutError:
        request_task.cancel()
        await asyncio.gather(request_task, return_exceptions=True)
        raise ErpcTimeoutError(timeout) from None
    finally:
        if cancel_task is not None and not cancel_task.done():
            cancel_task.cancel()


def _valid_id(value: object) -> bool:
    return (isinstance(value, int) and not isinstance(value, bool)) or isinstance(value, str)


class HttpJsonRpcTransport:
    """Credential-safe transport for intact JSON-RPC requests and batches."""

    def __init__(
        self,
        *,
        api_key: str | None,
        endpoint: str,
        headers: Mapping[str, str],
        timeout: float,
        client: httpx.AsyncClient,
        max_batch_size: int = 256,
        direct: bool = False,
        redactions: Iterable[str] = (),
    ) -> None:
        self._api_key = api_key
        self._endpoint = endpoint
        self._headers = dict(headers)
        self._timeout = timeout
        self._client = client
        self._max_batch_size = max_batch_size
        self._direct = direct
        self._redactions = tuple(redactions)
        self._ids = itertools.count(1)

    @property
    def endpoint(self) -> str:
        """Return the public endpoint without credentials."""
        return strip_url_query(self._endpoint) if self._direct else self._endpoint

    @property
    def max_batch_size(self) -> int:
        return self._max_batch_size

    def _id(self) -> int:
        return next(self._ids)

    async def request(
        self,
        method: str,
        params: JsonRpcParams | None = None,
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> JsonValue:
        request_id = self._id()
        body: dict[str, object] = {
            "jsonrpc": "2.0",
            "id": request_id,
            "method": method,
        }
        if params is not None:
            body["params"] = params
        response = await self._post(body, options)
        return self._unwrap(response, request_id)

    async def batch(
        self,
        calls: Sequence[RpcBatchCall],
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> list[JsonValue]:
        if not calls:
            return []
        if len(calls) > self._max_batch_size:
            raise ErpcInvalidResponseError(
                f"A batch may contain at most {self._max_batch_size} calls"
            )
        requests: list[dict[str, object]] = []
        for call in calls:
            item: dict[str, object] = {
                "jsonrpc": "2.0",
                "id": self._id(),
                "method": call["method"],
            }
            if "params" in call:
                item["params"] = call["params"]
            requests.append(item)
        response = await self._post(requests, options)
        if not isinstance(response, list):
            if isinstance(response, dict) and isinstance(response.get("error"), dict):
                self._raise_rpc(cast(dict[str, object], response["error"]))
            raise ErpcInvalidResponseError("ERPC returned a non-array batch response")
        by_id: dict[JsonRpcId, object] = {}
        for response_item in response:
            if not isinstance(response_item, dict) or not _valid_id(response_item.get("id")):
                raise ErpcInvalidResponseError("ERPC returned an invalid batch item")
            item_id = cast(JsonRpcId, response_item["id"])
            if item_id in by_id:
                raise ErpcInvalidResponseError("ERPC returned a duplicate batch id")
            by_id[item_id] = response_item
        results: list[JsonValue] = []
        for request in requests:
            request_id = cast(JsonRpcId, request["id"])
            response_item = by_id.pop(request_id, None)
            if response_item is None:
                raise ErpcInvalidResponseError(f"ERPC omitted batch response id {request_id}")
            results.append(self._unwrap(response_item, request_id))
        if by_id:
            raise ErpcInvalidResponseError("ERPC returned an unexpected batch response id")
        return results

    async def _post(self, body: object, options: RequestOptions) -> object:
        transport_error: ErpcTransportError | None = None
        try:
            if self._direct:
                request = _direct_request(
                    self._endpoint,
                    _direct_headers(self._headers),
                    body,
                )
                response = await _wait(
                    self._client.send(
                        request,
                        auth=None,
                        follow_redirects=False,
                    ),
                    self._timeout,
                    options,
                )
            else:
                if self._api_key is None:
                    raise ErpcNotConfiguredError("rpc")
                response = await _wait(
                    self._client.post(
                        self._endpoint,
                        params={"api-key": self._api_key},
                        headers=self._headers,
                        json=body,
                    ),
                    self._timeout,
                    options,
                )
        except (ErpcAbortedError, ErpcTimeoutError):
            raise
        except httpx.HTTPError:
            transport_error = ErpcTransportError()
        if transport_error is not None:
            # Raise after leaving the native exception handler so its context
            # is not retained on the SDK error object.
            raise transport_error
        if not response.is_success:
            raise ErpcHttpError(response.status_code)
        return _response_json(response)

    def _unwrap(self, response: object, expected_id: JsonRpcId) -> JsonValue:
        if not isinstance(response, dict) or response.get("id") != expected_id:
            raise ErpcInvalidResponseError("ERPC returned an unexpected response id")
        if "error" in response:
            error = response["error"]
            if not isinstance(error, dict):
                raise ErpcInvalidResponseError("ERPC returned an invalid RPC error")
            self._raise_rpc(cast(dict[str, object], error))
        if "result" not in response:
            raise ErpcInvalidResponseError("ERPC returned an invalid response")
        return cast(JsonValue, response["result"])

    def _raise_rpc(self, error: dict[str, object]) -> None:
        code = error.get("code")
        message = error.get("message")
        if not isinstance(code, int) or isinstance(code, bool) or not isinstance(message, str):
            raise ErpcInvalidResponseError("ERPC returned an invalid RPC error")
        data = error.get("data")
        credentials = self._redactions if self._direct else (self._api_key or "")
        raise ErpcJsonRpcError(
            code,
            redact_text(message, credentials),
            redact_value(data, credentials) if data is not None else None,
        )


class UnavailableHttpJsonRpcTransport(HttpJsonRpcTransport):
    """A local transport target for an unconfigured RPC namespace."""

    def __init__(self, namespace: str) -> None:
        self._namespace = namespace
        self._max_batch_size = 256

    @property
    def endpoint(self) -> str:
        return ""

    @property
    def max_batch_size(self) -> int:
        return self._max_batch_size

    async def request(
        self,
        method: str,
        params: JsonRpcParams | None = None,
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> JsonValue:
        del method, params, options
        raise ErpcNotConfiguredError(self._namespace)

    async def batch(
        self,
        calls: Sequence[RpcBatchCall],
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> list[JsonValue]:
        del options
        if not calls:
            return []
        if len(calls) > self._max_batch_size:
            raise ErpcInvalidResponseError(
                f"A batch may contain at most {self._max_batch_size} calls"
            )
        raise ErpcNotConfiguredError(self._namespace)


class RestTransport:
    """Bearer-authenticated async JSON and SSE transport."""

    def __init__(
        self,
        *,
        credential: str,
        endpoint: str,
        headers: Mapping[str, str],
        timeout: float,
        client: httpx.AsyncClient,
    ) -> None:
        self._credential = credential
        self._endpoint = endpoint
        self._headers = dict(headers)
        self._timeout = timeout
        self._client = client

    @property
    def endpoint(self) -> str:
        return self._endpoint

    def url(self, path: str) -> str:
        return endpoint_with_path(self._endpoint, path)

    @staticmethod
    def query(query: Mapping[str, QueryValue] | None) -> httpx.QueryParams:
        result = httpx.QueryParams()
        for key, value in (query or {}).items():
            if value is None:
                continue
            if isinstance(value, Sequence) and not isinstance(value, str):
                for item in value:
                    result = result.add(key, item)
            elif isinstance(value, bool):
                result = result.add(key, str(value).lower())
            else:
                result = result.add(key, str(value))
        return result

    def _request(self, path: str, query: Mapping[str, QueryValue] | None) -> httpx.Request:
        headers = {
            **self._headers,
            "authorization": f"Bearer {self._credential}",
            "accept": "application/json",
        }
        return self._client.build_request(
            "GET",
            self.url(path),
            params=self.query(query),
            headers=headers,
        )

    async def get(
        self,
        path: str,
        query: Mapping[str, QueryValue] | None = None,
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> JsonValue:
        try:
            response = await _wait(
                self._client.send(self._request(path, query)), self._timeout, options
            )
        except (ErpcAbortedError, ErpcTimeoutError):
            raise
        except httpx.HTTPError:
            raise ErpcTransportError() from None
        if not response.is_success:
            raise ErpcHttpError(response.status_code)
        return cast(JsonValue, _response_json(response))

    @asynccontextmanager
    async def stream(
        self,
        path: str,
        query: Mapping[str, QueryValue] | None = None,
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> AsyncIterator[httpx.Response]:
        request = self._request(path, query)
        try:
            response = await _wait(self._client.send(request, stream=True), self._timeout, options)
        except (ErpcAbortedError, ErpcTimeoutError):
            raise
        except httpx.HTTPError:
            raise ErpcTransportError() from None
        if not response.is_success:
            await response.aclose()
            raise ErpcHttpError(response.status_code)
        try:
            yield response
        except httpx.HTTPError:
            raise ErpcTransportError() from None
        finally:
            await response.aclose()


class UnavailableRestTransport(RestTransport):
    """A local transport target for an unconfigured REST/SSE namespace."""

    def __init__(self, namespace: str) -> None:
        self._namespace = namespace

    @property
    def endpoint(self) -> str:
        return ""

    def url(self, path: str) -> str:
        del path
        return ""

    async def get(
        self,
        path: str,
        query: Mapping[str, QueryValue] | None = None,
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> JsonValue:
        del path, query, options
        raise ErpcNotConfiguredError(self._namespace)

    @asynccontextmanager
    async def stream(
        self,
        path: str,
        query: Mapping[str, QueryValue] | None = None,
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> AsyncIterator[httpx.Response]:
        del path, query, options
        raise ErpcNotConfiguredError(self._namespace)
        yield cast(httpx.Response, None)


def strip_url_query(url: str) -> str:
    """Return a URL safe for public display."""
    parsed = urlsplit(url)
    return urlunsplit((parsed.scheme, parsed.netloc, parsed.path, "", ""))

File: packages/ruby/lib/erpc_sdk/transport.rb
# frozen_string_literal: true

require "json"
require "net/http"
require "openssl"
require "socket"
require "thread"
require "uri"

module ERPC
  HttpResponse = Struct.new(:status, :body, keyword_init: true)

  class NetHttpAdapter
    def request(method:, url:, headers:, body: nil, timeout: DEFAULT_TIMEOUT)
      uri = URI.parse(url)
      request = request_class(method).new(uri.request_uri, headers)
      request.body = body unless body.nil?
      response = start(uri, timeout) { |http| http.request(request) }
      HttpResponse.new(status: response.code.to_i, body: response.body.to_s)
    rescue Net::OpenTimeout, Net::ReadTimeout, Net::WriteTimeout
      raise TimeoutError, timeout, cause: nil
    rescue IOError, EOFError, SocketError, SystemCallError, OpenSSL::SSL::SSLError
      raise TransportError, "Unable to reach ERPC", cause: nil
    end

    def stream(url:, headers:, timeout: DEFAULT_TIMEOUT)
      Enumerator.new do |yielder|
        uri = URI.parse(url)
        request = Net::HTTP::Get.new(uri.request_uri, headers)
        start(uri, timeout) do |http|
          http.request(request) do |response|
            status = response.code.to_i
            raise HttpError, status unless status.between?(200, 299)

            response.read_body { |chunk| yielder << chunk }
          end
        end
      rescue Net::OpenTimeout, Net::ReadTimeout, Net::WriteTimeout
        raise TimeoutError, timeout, cause: nil
      rescue HttpError
        raise
      rescue IOError, EOFError, SocketError, SystemCallError, OpenSSL::SSL::SSLError
        raise TransportError, "Unable to reach ERPC", cause: nil
      end
    end

    private

    def start(uri, timeout, &block)
      Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https", open_timeout: timeout,
                      read_timeout: timeout, write_timeout: timeout, &block)
    end

    def request_class(method)
      { get: Net::HTTP::Get, post: Net::HTTP::Post }.fetch(method.to_sym)
    end
  end

  class HttpJsonRpcTransport
    attr_reader :max_batch_size

    def initialize(api_key: nil, endpoint:, headers: {}, timeout:, adapter:, max_batch_size: 256,
                   direct: false, redactions: [], unavailable_namespace: nil)
      @api_key = api_key
      @endpoint = endpoint
      @public_endpoint = URLs.public_endpoint(endpoint)
      @headers = headers.to_h.dup.freeze
      @timeout = timeout
      @adapter = adapter
      @max_batch_size = max_batch_size
      @direct = direct
      @redactions = redactions
      @unavailable_namespace = unavailable_namespace
      @next_id = 0
      @id_mutex = Mutex.new
    end

    def endpoint
      @public_endpoint
    end

    def inspect
      "#<#{self.class} endpoint=#{endpoint.inspect} max_batch_size=#{max_batch_size.inspect}>"
    end

    def request(method, params = nil)
      ensure_configured
      request_id = next_id
      body = { "jsonrpc" => "2.0", "id" => request_id, "method" => method }
      body["params"] = params unless params.nil?
      unwrap(post(body), request_id)
    end

    def batch(calls)
      return [] if calls.empty?
      if calls.length > max_batch_size
        raise InvalidResponseError, "A batch may contain at most #{max_batch_size} calls"
      end
      ensure_configured

      requests = calls.map do |call|
        method = call.fetch(:method) { call.fetch("method") }
        item = { "jsonrpc" => "2.0", "id" => next_id, "method" => method }
        if call.key?(:params) || call.key?("params")
          item["params"] = call.key?(:params) ? call[:params] : call["params"]
        end
        item
      rescue KeyError
        raise ConfigError, "batch calls require a method"
      end

      response = post(requests)
      unless response.is_a?(Array)
        raise_rpc(response["error"]) if response.is_a?(Hash) && response["error"].is_a?(Hash)
        raise InvalidResponseError, "ERPC returned a non-array batch response"
      end

      by_id = {}
      response.each do |item|
        unless item.is_a?(Hash) && valid_id?(item["id"])
          raise InvalidResponseError, "ERPC returned an invalid batch item"
        end
        raise InvalidResponseError, "ERPC returned a duplicate batch id" if by_id.key?(item["id"])

        by_id[item["id"]] = item
      end

      results = requests.map do |request|
        item = by_id.delete(request["id"])
        raise InvalidResponseError, "ERPC omitted a batch response" unless item

        unwrap(item, request["id"])
      end
      raise InvalidResponseError, "ERPC returned an unexpected batch response id" unless by_id.empty?

      results
    end

    private

    def ensure_configured
      return if @unavailable_namespace.nil?

      raise NotConfiguredError, @unavailable_namespace
    end

    def next_id
      @id_mutex.synchronize { @next_id += 1 }
    end

    def post(body)
      ensure_configured
      uri = URI.parse(@endpoint)
      unless @direct
        query = URI.decode_www_form(uri.query.to_s)
        query << ["api-key", @api_key]
        uri.query = URI.encode_www_form(query)
      end
      response = @adapter.request(
        method: :post,
        url: uri.to_s,
        headers: protocol_headers,
        body: JSON.generate(body),
        timeout: @timeout
      )
      raise HttpError, response.status unless response.status.between?(200, 299)

      JSON.parse(response.body)
    rescue JSON::ParserError
      raise InvalidResponseError, "ERPC returned malformed JSON", cause: nil
    end

    def unwrap(response, expected_id)
      unless response.is_a?(Hash) && response["id"] == expected_id
        raise InvalidResponseError, "ERPC returned an unexpected response id"
      end
      raise_rpc(response["error"]) if response.key?("error")
      raise InvalidResponseError, "ERPC returned an invalid response" unless response.key?("result")

      response["result"]
    end

    def raise_rpc(error)
      unless error.is_a?(Hash) && error["code"].is_a?(Integer) && error["message"].is_a?(String)
        raise InvalidResponseError, "ERPC returned an invalid RPC error"
      end

      credentials = @direct ? @redactions : @api_key
      data = error.key?("data") ? Redaction.value(error["data"], credentials) : nil
      raise JsonRpcError.new(error["code"], Redaction.text(error["message"], credentials), data)
    end

    def protocol_headers
      headers = @headers.dup
      if @direct
        headers.delete_if { |name, _| %w[accept content-type].include?(name.to_s.downcase) }
      end
      headers.merge("accept" => "application/json", "content-type" => "application/json")
    end

    def valid_id?(value)
      value.is_a?(Integer) || value.is_a?(String)
    end
  end

  class RestTransport
    def initialize(credential: nil, endpoint:, headers: {}, timeout:, adapter:, unavailable_namespace: nil)
      @credential = credential
      @endpoint = endpoint
      @public_endpoint = URLs.public_endpoint(endpoint)
      @headers = headers.to_h.dup.freeze
      @timeout = timeout
      @adapter = adapter
      @unavailable_namespace = unavailable_namespace
    end

    def endpoint
      @public_endpoint
    end

    def inspect
      "#<#{self.class} endpoint=#{endpoint.inspect}>"
    end

    def get(path, query = nil)
      ensure_configured
      response = @adapter.request(
        method: :get,
        url: url(path, query),
        headers: @headers.merge("authorization" => "Bearer #{@credential}", "accept" => "application/json"),
        timeout: @timeout
      )
      raise HttpError, response.status unless response.status.between?(200, 299)

      JSON.parse(response.body)
    rescue JSON::ParserError
      raise InvalidResponseError, "ERPC returned malformed JSON", cause: nil
    end

    def stream(path, query = nil)
      ensure_configured
      @adapter.stream(
        url: url(path, query),
        headers: @headers.merge("authorization" => "Bearer #{@credential}", "accept" => "text/event-stream"),
        timeout: @timeout
      )
    end

    private

    def ensure_configured
      return if @unavailable_namespace.nil?

      raise NotConfiguredError, @unavailable_namespace
    end

    def url(path, query)
      uri = URI.parse(URLs.with_path(endpoint, path))
      pairs = []
      (query || {}).each do |key, value|
        next if value.nil?

        if value.is_a?(Array)
          value.each { |item| pairs << [key.to_s, item.to_s] }
        else
          rendered = value == true ? "true" : value == false ? "false" : value.to_s
          pairs << [key.to_s, rendered]
        end
      end
      uri.query = URI.encode_www_form(pairs) unless pairs.empty?
      uri.to_s
    end
  end
end
